proposals/004_Overlays.md

```diff
@@ -0,0 +1,194 @@
+
+```yaml
+overlay: 1.0.0
+info:
+  title: Structured Overlay
+  version: 1.0.0
+updates:
+- target: paths."/foo".get
+    value:
+        description: This is the new description
+- target: paths."/bar".get
+    value:
+        description: This is the updated description
+- target: paths."/bar"
+    value:
+        post:
+            description: This is an updated description of a child object
+            x-safe: false
+```
+
+##### Wildcard Overlays Examples
+
+One significant advantage of using the JMESPath syntax that it allows referencing multiple nodes in the target document.  This would allow a single update object to be applied to multiple target objects using wildcards.
+
+```yaml
+overlay: 1.0.0
+info:
+  title: Update many objects at once
+  version: 1.0.0
+updates:
+- target: paths.*.get
+    value:
+      x-safe: true
+- target: paths.*.get.parameters[?name=='filter' && in=='query']
+    value:
+      schema:
+        $ref: "/components/schemas/filterSchema"
+```
+
+##### Array Modification Examples
+
+Due to the fact that we can now reference specific elements of the parameter array, it allows adding parameters. Parameters can be deleted using the `remove` property.  Use of indexes to remove array items should be avoided where possible as indexes will change when items are removed.
+
+```yaml
+overlay: 1.0.0
+info:
+  title: Add an array element
+  version: 1.0.0
+updates:
+- target: paths.*.get.parameters[length(@)]
+    value: 
+      name: newParam
+      in: query
+```
+
+```yaml
+overlay: 1.0.0
+info:
+  title: Remove a array element
+  version: 1.0.0
+updates:
+- target: $.paths[*].get.parameters[? name == 'dummy']
+    remove: true
+```
+
+## Proposal Summary
+
+### Benefits
+
+- This approach addresses the two distinct approaches of structured overlay vs targeted overlay which suits distinct but equally valid scenarios.
+- Addresses the problem of modifying the parameters array and removes the need to replace the entire array when a small change is required.
+- Allows sets of related overlays to be stored in a same file.
+- Enables updating a set of objects based on a pattern. This might be an effective way of apply common behaviour across many operations in an API.
+
+### Challenges
+
+- Tooling will need a JMESPath implementation.
+- Large overlays may be slow to process.
+- Multiple complex pattern based overlays may cause overlapping updates causing confusing outcomes.
+
+## Alternatives considered
+
+JMESPath was chosen over JSONPath due to the fact that JMESPath has a [specification](http://jmespath.org/specification.html) and a set of test cases.  This will help to ensure compatibility between implementations.
+
+## Backwards compatibility
+
+Overlays will be described in a new specification that can be used alongside an OpenAPI Description, therefore there will be no compatibility issues for the initial release. Overlay documents can be used against OpenAPI v2 and v3 descriptions.
```

Comment:
I'm glad to see this, though it might be good to only support v3 to help give people another reason to move to v3.

Reply:
I considered it.  I just didn't like the idea of creating an arbitrary constraint that I couldn't really justify.  I'm on the fence.  I suspect if we tell people they can't people will do it anyway.